chalice/local.py

```diff
@@ -214,29 +434,106 @@ def _send_autogen_options_response(self, lambda_event):
         # So our local version needs to add this manually to our set of allowed
         # headers.
         route_methods.append('OPTIONS')
-        route_methods = route_methods
 
         # The Access-Control-Allow-Methods header is not added by the
         # CORSConfig object it is added to the API Gateway route during
         # deployment, so we need to manually add those headers here.
         cors_headers.update({
             'Access-Control-Allow-Methods': '%s' % ','.join(route_methods)
         })
+        return cors_headers
+
+    def _handle_binary(self, response):
+        # type: (Dict[str,Any]) -> Dict[str,Any]
+        if response.get('isBase64Encoded'):
+            body = base64.b64decode(response['body'])
+            response['body'] = body
+        return response
+
+
+class ChaliceRequestHandler(BaseHTTPRequestHandler):
+    """A class for mapping raw HTTP events to and from LocalGateway."""
+    protocol_version = 'HTTP/1.1'
+
+    def __init__(self, request, client_address, server, app_object, config):
+        # type: (bytes, Tuple[str, int], HTTPServer, Chalice, Config) -> None
+        self.local_gateway = LocalGateway(app_object, config)
+        BaseHTTPRequestHandler.__init__(
+            self, request, client_address, server)  # type: ignore
+
+    def _parse_payload(self):
+        # type: () -> Tuple[HeaderType, str]
+        body = None
+        content_length = int(self.headers.get('content-length', '0'))
+        if content_length > 0:
+            body = self.rfile.read(content_length)
+        # mypy doesn't like dict(self.headers) so I had to use a
+        # dictcomp instead to make it happy.
+        converted_headers = {key: value for key, value in self.headers.items()}
+        return converted_headers, body
+
+    def _generic_handle(self):
+        # type: () -> None
+        headers, body = self._parse_payload()
+        try:
+            response = self.local_gateway.handle_request(
+                method=self.command,
+                path=self.path,
+                headers=headers,
+                body=body
+            )
+            self._send_http_response(response)
+        except NotAuthorizedError as e:
+            self._send_not_authorized_response(e.headers)
+        except NoOptionsRouteDefined as e:
+            self._send_http_response_no_body(e.headers)
+
+    def _send_not_authorized_response(self, headers):
+        # type: (HeaderType) -> None
+        self.send_response(401)
```

Comment:
An authorized error where we generate a policy that forbids access to a resource gives a 403, not a 401.

```
$ http https://rest-api-id.execute-api.us-west-2.amazonaws.com/api/ 'Authorization: deny'
HTTP/1.1 403 Forbidden
Connection: keep-alive
Content-Length: 60
Content-Type: application/json
Date: Thu, 24 Aug 2017 00:27:57 GMT
Via: 1.1 7f3f42df8af148df1f9f1ee7175987ad.cloudfront.net (CloudFront)
X-Amz-Cf-Id: 4ZhC3GokHeXr68EKakpF6xQWxVvTDZtQiGK253psOcWbPW3M980rGw==
X-Cache: Error from cloudfront
x-amzn-ErrorType: AccessDeniedException
x-amzn-RequestId: 13ff2f6d-8863-11e7-af78-e1cc9bcdaa69

{
    "Message": "User is not authorized to access this resource"
}
```